main/src/ca/uwaterloo/flix/language/phase/Parser2.scala

```diff
@@ -639,11 +639,48 @@ object Parser2 {
   private val NAME_PREDICATE: Set[TokenKind] = Set(TokenKind.NameUpperCase)
 
   /**
-    * Consumes a token if kind is in `kinds`.
-    * If `allowQualified` is passed also consume subsequent dot-separated tokens with kind in `kinds`.
-    * If the found kind is in the tail set, then no further dot separated tokens will be consume
+    *   - `KIND` is consumed and produces [[TreeKind.Ident]]
+    *   - [[TokenKind.isFirstDecl]] and [[TokenKind.isKeyword]] (not consumed)
+    *     produces [[TreeKind.ErrorTree]]
+    *   - `ANY` is consumed and produces [[TreeKind.ErrorTree]]
     */
-  private def name(kinds: Set[TokenKind], tail: Set[TokenKind] = Set(TokenKind.NameLowerCase), allowQualified: Boolean = false, context: SyntacticContext)(implicit s: State): Mark.Closed = {
+  private def name(kinds: Set[TokenKind], context: SyntacticContext)(implicit s: State): Mark.Closed = {
+    nameGeneric(kinds, tail = Set.empty, allowQualified = false, context)
+  }
+
+  /**
+    *   - `KIND` is consumed and produces [[TreeKind.Ident]]
+    *   - `KIND.KIND.KIND` is consumed and produces [[TreeKind.QName]]([[TreeKind.Ident]]..)
+    *   - `KIND.KIND.` is consumed and produces
+    *     [[TreeKind.QName]]([[TreeKind.Ident]].., [[TreeKind.ErrorTree]])
+    *   - `KIND. KIND.KIND` is consumed and produces
+    *     [[TreeKind.QName]]([[TreeKind.Ident]].., [[TreeKind.ErrorTree]], [[TreeKind.Ident]]..)
+    *   - [[TokenKind.isFirstDecl]] and [[TokenKind.isKeyword]] (not consumed)
+    *     produces [[TreeKind.ErrorTree]]
+    *   - `ANY` is consumed and produces [[TreeKind.ErrorTree]]
+    *
+    * @param tail if any KIND found is in `tail` then no more dots are parsed
+    */
+  private def qualifiedName(kinds: Set[TokenKind], tail: Set[TokenKind] = Set(TokenKind.NameLowerCase), context: SyntacticContext)(implicit s: State): Mark.Closed = {
+    nameGeneric(kinds, tail, allowQualified = true, context)
+  }
+
+  /**
+    *   - `KIND` is consumed and produces [[TreeKind.Ident]]
+    *   - `KIND.KIND.KIND` is consumed and produces [[TreeKind.QName]]([[TreeKind.Ident]]..)
+    *   - `KIND.KIND.` is consumed and produces
+    *     [[TreeKind.QName]]([[TreeKind.Ident]].., [[TreeKind.ErrorTree]])
+    *   - `KIND. KIND.KIND` is consumed and produces
+    *     [[TreeKind.QName]]([[TreeKind.Ident]].., [[TreeKind.ErrorTree]], [[TreeKind.Ident]]..)
+    *   - [[TokenKind.isFirstDecl]] and [[TokenKind.isKeyword]] (not consumed)
+    *     produces [[TreeKind.ErrorTree]]
+    *   - `ANY` is consumed and produces [[TreeKind.ErrorTree]]
```

Comment:
I am not sure how to do these comments without filling the screen, feels like i need a parsing rule specification language